### ISSUE TITLE
Colab giving different results to book for Chapter 3

### ISSUE THREAD

**Issue by Bazman76** [ageron/handson-ml2#197]
The results I get when running your Colab file for Chapter 3 are different to the results in the book.

The first divergence comes from the results in blocks 17&18 of the Colab file, which corresponds to p111-113 of the book, when run on my machine both blocks give the same result:

array([0.95035, 0.96035, 0.9604])

This is of course different to the results shown in the book ( I realise the differences are not huge but since we are setting the seed for the random numbers they should be the exactly the same?)

I note that block 20 does give the exact same answer as in the book.

The confusion matrix produced by block 22 is also different to that shown in the book.

array([[53892,   687],
       [ 1891,  3530]])

As you might expect all the associated precision, recall and f1 scores are also different to the book.  

I also attach the graphs produced by blocks 36&38:

As you can see in the plot "HOML1" produced by block 36 the Precision is much bumpier than the book version.

The graph produced by block 38 "HOML2" is also different with the one produced on my machine showing a large initial spike downward in Precision from 1 to 0.8. There is a much smaller initial spike downwards in the book version.

![HOML1](https://user-images.githubusercontent.com/9517425/86248691-1f297f80-bba6-11ea-9a9d-451f90d3f4b0.png)
![HOML2](https://user-images.githubusercontent.com/9517425/86248700-22247000-bba6-11ea-8255-4491355870af.png)

Lastly, this is as far as I went, the value I get for threshold_90_precision in block 40:

3370.0194991439557

Which is very different to the 7816 shown in the book p117-120.

Can you please advise what I need to do to match the results in the book?

**Comment by Bazman76** [comment]
This seems to answer some of my issues: https://github.com/ageron/handson-ml2/issues/152, sorry I did not see it earlier. 
Are the issues with cells 17&18 also caused by different sklearn versions?
How about cells 38 and 40?

**Comment by rickiepark** [comment]
Hi,

Yes, It's caused by sklearn version including cell 38 and 40.
If you install sklearn 0.20.4, you can see the same result in the book.
Because some bugs in SGD is fixed at sklearn 0.21.0 (https://github.com/scikit-learn/scikit-learn/pull/13422).

Cheers,
Haesun

**Comment by Bazman76** [comment]
That's great thanks for explaining!

**Comment by GuidoMr** [comment]
Hello, I also had this same problems but since it wasn't a big deal I didn't think much of it. However, on page 102 of the printed book, on the "Multiclass Classification" part, when I run 
`sgd_clf.fit(X_train, y_train)
sgd_clf.predict([some_digit])`
I get this output:
`array([3], dtype=uint8)`
different than the
`array([5], dtype=uint8)` shown in the book.
When I run `sgd_clf.decision_function([some_digit])` I get this output: `array([[-31893.03095419, -34419.69069632,  -9530.63950739,
          1823.73154031, -22320.14822878,  -1385.80478895,
        -26188.91070951, -16147.51323997,  -4604.35491274,
        -12050.767298  ]])`. This is definitely a big deal since those results are completely different from the ones shown in the book, different than the others. Did you have this same problem and could you fix it without changing the sklearn version? Thanks in advance!
